rest_framework/decorators.py

```diff
@@ -130,29 +131,51 @@ def decorator(func):
     return decorator
 
 
-def detail_route(methods=None, **kwargs):
+def action(methods=None, detail=True, name=None, url_name=None, url_path=None, **kwargs):
     """
-    Used to mark a method on a ViewSet that should be routed for detail requests.
+    Mark a ViewSet method as a routable action.
+
+    Set the `detail` boolean to determine if this action should apply to
+    instance/detail requests or collection/list requests.
     """
     methods = ['get'] if (methods is None) else methods
+    methods = [method.lower() for method in methods]
 
     def decorator(func):
         func.bind_to_methods = methods
-        func.detail = True
+        func.detail = detail
+        func.name = name or pretty_name(func.__name__)
+        func.url_name = url_name or func.__name__
+        func.url_path = url_path or func.__name__
         func.kwargs = kwargs
+        func.kwargs.update({
+            'name': func.name,
+            'description': func.__doc__ or None
+        })
+
         return func
     return decorator
 
 
+def detail_route(methods=None, **kwargs):
+    """
+    Used to mark a method on a ViewSet that should be routed for detail requests.
+    """
+    warnings.warn(
+        "`detail_route` has been deprecated in favor of `action`, which accepts a "
+        "`detail` boolean. Use `@action(['GET'], detail=True)` instead.",
+        PendingDeprecationWarning
```

Comment:
I think this needs a `stacklevel` parameter so that the source code in the traceback points to the file using detail_route, not this file.